ray-operator/controllers/ray/rayjob_controller.go

```diff
@@ -192,40 +194,46 @@ func (r *RayJobReconciler) Reconcile(ctx context.Context, request ctrl.Request)
 		return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, err
 	}
 
-	// Check the current status of ray jobs before submitting.
-	jobInfo, err := rayDashboardClient.GetJobInfo(ctx, rayJobInstance.Status.JobId)
+	// Ensure k8s job has been created
+	jobName, err := r.getOrCreateK8sJob(ctx, rayJobInstance)
 	if err != nil {
-		err = r.updateState(ctx, rayJobInstance, jobInfo, rayJobInstance.Status.JobStatus, rayv1alpha1.JobDeploymentStatusFailedToGetJobStatus, err)
-		// Dashboard service in head pod takes time to start, it's possible we get connection refused error.
-		// Requeue after few seconds to avoid continuous connection errors.
+		r.Log.Error(err, "failed to create k8s job")
+		err = r.updateState(ctx, rayJobInstance, nil, rayJobInstance.Status.JobStatus, rayv1alpha1.JobDeploymentStatusFailedJobDeploy, err)
 		return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, err
 	}
 
-	r.Log.V(1).Info("RayJob information", "RayJob", rayJobInstance.Name, "jobInfo", jobInfo, "rayJobInstance", rayJobInstance.Status.JobStatus)
-	if jobInfo == nil {
-		// Submit the job if no id set
-		jobId, err := rayDashboardClient.SubmitJob(ctx, rayJobInstance, &r.Log)
-		if err != nil {
-			r.Log.Error(err, "failed to submit job")
-			err = r.updateState(ctx, rayJobInstance, jobInfo, rayJobInstance.Status.JobStatus, rayv1alpha1.JobDeploymentStatusFailedJobDeploy, err)
-			return ctrl.Result{}, err
-		}
+	r.Log.Info("K8s job successfully created", "RayJob", rayJobInstance.Name, "jobId", jobName)
+	r.Recorder.Eventf(rayJobInstance, corev1.EventTypeNormal, "Created", "Created k8s job %s", jobName)
 
-		r.Log.Info("Job successfully submitted", "RayJob", rayJobInstance.Name, "jobId", jobId)
-		r.Recorder.Eventf(rayJobInstance, corev1.EventTypeNormal, "Submitted", "Submit Job %s", jobId)
-		// Here, we directly update to PENDING and emit an event to trigger a new reconcile loop
-		err = r.updateState(ctx, rayJobInstance, jobInfo, rayv1alpha1.JobStatusPending, rayv1alpha1.JobDeploymentStatusRunning, nil)
-		if err != nil {
+	// Check the status of the k8s job and update the RayJobInstance status accordingly.
+	// Get the k8s job
+	k8sJob := &batchv1.Job{}
+	err = r.Client.Get(ctx, types.NamespacedName{Name: jobName, Namespace: rayJobInstance.Namespace}, k8sJob)
```

Comment:
Good point, I will optimize this

Reply:
Is it safer to call `Get()` again to get the latest status, or does it not matter because the time in between the two "Get" operations is short enough? I'm not sure of the performance overhead of calling `Get()` twice and if it's worth optimizing.

Reply:
We cannot guarantee that both Get() operations will have the same results, but the inconsistency should be handled by the Kubernetes concurrency model rather than the operator in my understanding.